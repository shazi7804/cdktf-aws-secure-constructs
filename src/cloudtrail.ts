import * as AWS from '@cdktf/provider-aws';
import { Resource } from 'cdktf';
import { Construct } from 'constructs';

export interface EnableCloudTrailProps {
  readonly bucketName?: string;
  readonly bucketKeyPrefix?: string;
  readonly tags?: { [key: string]: string };
}

export class EnableCloudTrail extends Resource {
  constructor(scope: Construct, name: string, props: EnableCloudTrailProps ) {
    super(scope, name);

    const region = new AWS.DataAwsRegion(this, 'currentRegion').name;
    const accountId = new AWS.DataAwsCallerIdentity(this, 'currentAccount').accountId;

    if (props.bucketName) {
      new AWS.Cloudtrail(this, 'EnableCloudTrail', {
        name: 'trail-secure-' + this,
        s3BucketName: props.bucketName,
      });
    } else {
      const bucket = new AWS.S3Bucket(this, 'VpcFlogLogBucket', {
        bucket: 'cloudtrail-secure-' + region + '-' + accountId,
        tags: props.tags,
      });

      new AWS.Cloudtrail(this, 'EnableCloudTrail', {
        name: 'trail-secure-' + this,
        s3BucketName: bucket.id,
      });
    }

  }

  /**
   * Add Config Rule for Cloudtrail enabled
   * @param tags Config Rule tags
   */
  public addConfigRule(tags?: any): void {
    new AWS.ConfigConfigRule(this, 'EnableCloudTrailConfigRule', {
      name: 'CloudtrailEnabled',
      source: [{
        owner: 'AWS',
        sourceIdentifier: 'CLOUD_TRAIL_ENABLED',
      }],
      tags,
    });
  }
}
