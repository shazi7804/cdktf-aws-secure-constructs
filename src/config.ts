import * as AWS from '@cdktf/provider-aws';
import { Resource } from 'cdktf';
import { Construct } from 'constructs';

export interface EnableConfigRecorderProps {
    readonly snsTopicName: string;
    readonly recorderName?: string;
    readonly deliveryChannelName?: string;
    readonly recordingGroupAllSupported?: boolean;
    readonly bucketName?: string;
    readonly bucketKeyPrefix?: string;
    readonly tags?: { [key: string]: string };
  }

export class EnableConfigRecorder extends Resource {
    constructor(scope: Construct, name: string, props: EnableConfigRecorderProps ) {
    super(scope, name);

    const accountId = new AWS.DataAwsCallerIdentity(this, 'currentAccount').accountId;
    const region = new AWS.DataAwsRegion(this, 'currentRegion').name;

    const role = new AWS.IamRole(this, 'RecorderAssumeRole', {
        name: 'config-recorder-' + region,
        assumeRolePolicy: JSON.stringify({
            Version: "2012-10-17",
            Statement: [{
                Effect: "Allow",
                Principal: {
                    Service: "config.amazonaws.com"
                },
                Action: "sts:AssumeRole",
                Sid: ""
            }]
        })
    });

    new AWS.IamRolePolicyAttachment(this, 'RecorderAssumeRoleAttachment', {
        role: role.name as string,
        policyArn: 'arn:aws:iam::aws:policy/service-role/AWS_ConfigRole'
    });

    const recorder = new AWS.ConfigConfigurationRecorder(this, 'ConfigRemediationConfiguration', {
       name: props.recorderName || 'config-secure-' + accountId,
       roleArn: role.arn,
       recordingGroup: [{
           allSupported: props.recordingGroupAllSupported || true,
       }]
    });

    if (props.bucketName) {
        new AWS.ConfigDeliveryChannel(this, 'ConfigDeliveryChannel', {
            name: props.deliveryChannelName || 'config-secure-channel-' + accountId,
            s3BucketName: props.bucketName as string,
            s3KeyPrefix: props.bucketKeyPrefix,
            snsTopicArn: props.snsTopicName,
        })
    } else {
        const bucket = new AWS.S3Bucket(this, 'ConfigDeliveryChannelBucket', {
            bucket: 'config-secure-channel' + region + '-' + accountId,
            tags: props.tags,
        });

        new AWS.ConfigDeliveryChannel(this, 'ConfigDeliveryChannel', {
            name: props.deliveryChannelName || 'config-secure-channel-' + accountId,
            s3BucketName: bucket.id,
            s3KeyPrefix: props.bucketKeyPrefix,
            snsTopicArn: props.snsTopicName
        })
    }

    new AWS.ConfigConfigurationRecorderStatus(this, 'ConfigConfigurationRecorderStatus', {
        name: recorder.id,
        isEnabled: true
    })

  }
}
