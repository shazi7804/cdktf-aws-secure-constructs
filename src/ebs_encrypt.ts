import * as AWS from '@cdktf/provider-aws';
import { Resource } from 'cdktf';
import { Construct } from 'constructs';

export class EnableEbsEncryption extends Resource {
  constructor(scope: Construct, name: string ) {
    super(scope, name);

    new AWS.EbsEncryptionByDefault(this, 'EbsEncryptionByDefault', {
      enabled: true,
    });

  }

  /**
   * Add Config Rule for EBS default encrypt enabled
   * @param tags Config Rule tags
   */
  public addConfigRule(tags?: any): void {
    new AWS.ConfigConfigRule(this, 'EnableEbsEncryptionConfigRule', {
      name: 'EbsEncryptionEnabled',
      source: [{
        owner: 'AWS',
        sourceIdentifier: 'EC2_EBS_ENCRYPTION_BY_DEFAULT',
      }],
      tags,
    });
  }
}
