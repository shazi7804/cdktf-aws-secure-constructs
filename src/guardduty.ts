import * as AWS from '@cdktf/provider-aws';
import { Resource } from 'cdktf';
import { Construct } from 'constructs';

export interface EnableGuarddutyProps {
  readonly findingPublishingFrequency?: string;
  readonly members?: any;
  readonly tags?: { [key: string]: string };
}

export interface AcceptMemberAccountOptions {
  readonly accountId: string;
  readonly email: string;
}

export interface InviteAccepterMemberAccountOptions {
  readonly masterAccountId: string;
}

export class EnableGuardduty extends Resource {
  public readonly id: string;

  constructor(scope: Construct, name: string, props: EnableGuarddutyProps ) {
    super(scope, name);

    const gd = new AWS.GuarddutyDetector(this, 'DefaultGuarddutyDetector', {
      enable: true,
      findingPublishingFrequency: props.findingPublishingFrequency,
    });

    this.id = gd.id;
  }

  /**
   * Add Config Rule for Guardduty
   * @param tags Config Rule tags
   */
  public addConfigRule(tags?: any): void {
    new AWS.ConfigConfigRule(this, 'GuarddutyEnabledCertralizedConfigRule', {
      name: 'GuarddutyEnabledCertralized',
      source: [{
        owner: 'AWS',
        sourceIdentifier: 'GUARDDUTY_ENABLED_CENTRALIZED',
      }],
      tags,
    });
  }

  /**
   * Accept Member Account
   * @param id Guardduty member id
   * @param options GuarddutyMember options
   */
  public acceptMemberAccount(id: string, options: AcceptMemberAccountOptions): void {
    new AWS.GuarddutyMember(this, id, {
      detectorId: this.id,
      invite: true,
      invitationMessage: 'please accept guardduty invitation',

      accountId: options.accountId,
      email: options.email,
      disableEmailNotification: false,
    });
  }

  /**
   * Accept Member Account
   * @param id Guardduty member id
   * @param options GuarddutyMember options
   */
  public inviteAccepterMemberAccount(id: string, options: InviteAccepterMemberAccountOptions): void {
    new AWS.GuarddutyInviteAccepter(this, id, {
      detectorId: this.id,
      masterAccountId: options.masterAccountId,
    });
  }
}
