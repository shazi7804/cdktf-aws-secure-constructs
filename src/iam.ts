import * as AWS from '@cdktf/provider-aws';
import { Resource } from 'cdktf';
import { Construct } from 'constructs';

export interface CreateAccountPasswordPolicyProps {
  readonly minimumPasswordLength?: number;
  readonly passwordReusePrevention?: number;
  readonly requireNumbers?: boolean;
  readonly requireUppercaseCharacters?: boolean;
  readonly requireLowercaseCharacters?: boolean;
  readonly requireSymbols?: boolean;
  readonly allowUsersToChangePassword?: boolean;
  readonly maxPasswordAge?: number;
}

export class CreateAccountPasswordPolicy extends Resource {
  public readonly expirePasswords: boolean;

  constructor(scope: Construct, name: string, props: CreateAccountPasswordPolicyProps ) {
    super(scope, name);

    const policy = new AWS.IamAccountPasswordPolicy(this, 'defaultAccountPasswordPolicy', {
      minimumPasswordLength: props.minimumPasswordLength ?? 14,
      passwordReusePrevention: props.passwordReusePrevention ?? 24,
      requireNumbers: props.requireNumbers ?? true,
      requireUppercaseCharacters: props.requireUppercaseCharacters ?? true,
      requireLowercaseCharacters: props.requireLowercaseCharacters ?? true,
      requireSymbols: props.requireSymbols ?? true,
      allowUsersToChangePassword: props.allowUsersToChangePassword ?? true,
      maxPasswordAge: props.maxPasswordAge ?? 90,
    });

    this.expirePasswords = policy.expirePasswords;
  }

  /**
     * addConfigRule
     */
  public addConfigRule(tags: any): void {
    new AWS.ConfigConfigRule(this, 'defaultAccountPasswordPolicyConfigRule', {
      name: 'IAMAccountMFAEnabled',
      source: [{
        owner: 'AWS',
        sourceIdentifier: 'MFA_ENABLED_FOR_IAM_CONSOLE_ACCESS',
      }],
      tags,
    });
  }
}
