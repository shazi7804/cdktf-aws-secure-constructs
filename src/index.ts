export * from './iam';
export * from './vpc';
export * from './guardduty';
