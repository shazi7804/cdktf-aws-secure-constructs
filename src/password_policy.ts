import * as AWS from '@cdktf/provider-aws';
import { Resource } from 'cdktf';
import { Construct } from 'constructs';

export interface EnableAccountPasswordPolicyProps {
  readonly minimumPasswordLength?: number;
  readonly passwordReusePrevention?: number;
  readonly requireNumbers?: boolean;
  readonly requireUppercaseCharacters?: boolean;
  readonly requireLowercaseCharacters?: boolean;
  readonly requireSymbols?: boolean;
  readonly allowUsersToChangePassword?: boolean;
  readonly maxPasswordAge?: number;
}

export class EnableAccountPasswordPolicy extends Resource {
  public readonly expirePasswords: boolean;

  constructor(scope: Construct, name: string, props: EnableAccountPasswordPolicyProps ) {
    super(scope, name);

    const policy = new AWS.IamAccountPasswordPolicy(this, 'defaultAccountPasswordPolicy', {
      minimumPasswordLength: props.minimumPasswordLength ?? 14,
      passwordReusePrevention: props.passwordReusePrevention ?? 24,
      requireNumbers: props.requireNumbers ?? true,
      requireUppercaseCharacters: props.requireUppercaseCharacters ?? true,
      requireLowercaseCharacters: props.requireLowercaseCharacters ?? true,
      requireSymbols: props.requireSymbols ?? true,
      allowUsersToChangePassword: props.allowUsersToChangePassword ?? true,
      maxPasswordAge: props.maxPasswordAge ?? 90,
    });

    this.expirePasswords = policy.expirePasswords;
  }

  /**
   * Add Config Rule for Account Password policy
   * @param tags Config Rule tags
   */
  public addConfigRule(tags?: any): void {
    new AWS.ConfigConfigRule(this, 'defaultAccountPasswordPolicyConfigRule', {
      name: 'IAMAccountMFAEnabled',
      source: [{
        owner: 'AWS',
        sourceIdentifier: 'MFA_ENABLED_FOR_IAM_CONSOLE_ACCESS',
      }],
      tags,
    });
  }
}
