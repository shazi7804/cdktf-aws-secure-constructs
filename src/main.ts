import { Resource } from 'cdktf';
import { Construct } from 'constructs';
import { CreateAccountPasswordPolicy } from './iam';


export class Secure extends Resource {
    public readonly expirePasswords: boolean;

    constructor(scope: Construct, name: string ) {
        super(scope, name);

        new CreateAccountPasswordPolicy(this, 'def', {})
    }
}
