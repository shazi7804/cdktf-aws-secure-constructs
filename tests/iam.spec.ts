import { Testing, TerraformStack } from 'cdktf';
import { CreateAccountPasswordPolicy } from '../src';

describe('secure of IAM', () => {
    test('default account password policy', () => {
        const app = Testing.app();
        const stack = new TerraformStack(app, 'test');

        new CreateAccountPasswordPolicy(stack, 'TestDefault', {});

        expect(Testing.synth(stack)).toMatchSnapshot();
    });
});
