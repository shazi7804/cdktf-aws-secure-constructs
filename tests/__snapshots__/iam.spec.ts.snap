// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`secure of IAM 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_iam_account_password_policy\\": {
      \\"TestDefault_defaultAccountPasswordPolicy_DC7D980D\\": {
        \\"allow_users_to_change_password\\": true,
        \\"max_password_age\\": 90,
        \\"minimum_password_length\\": 14,
        \\"password_reuse_prevention\\": 24,
        \\"require_lowercase_characters\\": true,
        \\"require_numbers\\": true,
        \\"require_symbols\\": true,
        \\"require_uppercase_characters\\": true,
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/TestDefault/defaultAccountPasswordPolicy\\",
            \\"uniqueId\\": \\"TestDefault_defaultAccountPasswordPolicy_DC7D980D\\"
          }
        }
      }
    }
  }
}"
`;
